lib/maths/unittest/CBoostedTreeTest.cc

```diff
@@ -802,13 +818,225 @@ void CBoostedTreeTest::testDepthBasedRegularization() {
+                                      objective, 1e-3, maxIterations, expected,
+                                      objectiveAtExpected);
+            LOG_DEBUG(<< "expected = " << expected
+                      << " objective at expected = " << objectiveAtExpected);
+
+            CArgMinLogisticImpl argmin{lambda};
+            CArgMinLogisticImpl argminPartition[2]{{lambda}, {lambda}};
+            auto nextPass = [&] {
+                bool done{argmin.nextPass() == false};
+                done &= (argminPartition[0].nextPass() == false);
+                done &= (argminPartition[1].nextPass() == false);
+                return done == false;
+            };
+
+            do {
+                for (std::size_t i = 0; i < labels.size() / 2; ++i) {
+                    argmin.add(weights[i], labels[i]);
+                    argminPartition[0].add(weights[i], labels[i]);
+                }
+                for (std::size_t i = labels.size() / 2; i < labels.size(); ++i) {
+                    argmin.add(weights[i], labels[i]);
+                    argminPartition[1].add(weights[i], labels[i]);
+                }
+                argminPartition[0].merge(argminPartition[1]);
+                argminPartition[1] = argminPartition[0];
+            } while (nextPass());
+
+            double actual{argmin.value()};
+            double actualPartition{argminPartition[0].value()};
+            LOG_DEBUG(<< "actual = " << actual
+                      << " objective at actual = " << objective(actual));
+
+            // We should be within 1% for the value and 0.001% for the objective
+            // at the value.
+            CPPUNIT_ASSERT_EQUAL(actual, actualPartition);
+            CPPUNIT_ASSERT_DOUBLES_EQUAL(expected, actual, 0.01 * std::fabs(expected));
+            CPPUNIT_ASSERT_DOUBLES_EQUAL(objectiveAtExpected, objective(actual),
+                                         1e-5 * objectiveAtExpected);
+        }
+    }
+}
+
+void CBoostedTreeTest::testLogisticRegression() {
+
+    // Test we approximately minimise the cross entropy if the category labels
+    // are generated from log odds which are a linear function of the regressors.
+
+    test::CRandomNumbers rng;
+
+    std::size_t trainRows{1000};
+    std::size_t rows{1200};
+    std::size_t cols{4};
+    std::size_t capacity{600};
+
+    TMeanAccumulator meanExcessCrossEntropy;
+    for (std::size_t test = 0; test < 3; ++test) {
+        auto probability = [&] {
+            TDoubleVec weights;
+            rng.generateUniformSamples(-2.0, 2.0, cols - 1, weights);
+            TDoubleVec noise;
+            rng.generateNormalSamples(0.0, 1.0, rows, noise);
+            return [=](const TRowRef& row) {
+                double x{0.0};
+                for (std::size_t i = 0; i < cols - 1; ++i) {
+                    x += weights[i] * row[i];
+                }
+                return maths::CTools::logisticFunction(x + noise[row.index()]);
+            };
+        }();
```

Comment:
There isn't actually need in defining this as an anonymous lambda function, is there? Nesting of multiple lambda functions here hinders readability. Same also below.

Reply:
Done